R/sheets_speedread.R

```diff
@@ -0,0 +1,221 @@
+#' emphasis on the PDF format, which doesn't help us:
+#'   * <https://gist.github.com/Spencer-Easton/78f9867a691e549c9c70>
+#'
+#' @section The datasource URL:
+#'
+#' The "Chart Tools datasource protocol" is another way to access values in
+#' Sheets, designed to make Sheets play nicely with the Google Charts API. We
+#' can use the datasource URL as another way to get data as CSV (or JSON,
+#' etc.). It is more powerful than the export URL, in that it supports an
+#' entire SQL-like query language.
+#'
+#' The datasource URL does not appear to work with an Authorization header.
+#' Instead, we must send an access token as part of the query.
+#'
+#' Stack Overflow answer that links to various pages re: the "Chart Tools
+#' datasource protocol"
+#'   * <https://stackoverflow.com/a/33727897/2825349>
+#'
+#' The most important link is about "Query Source Ranges":
+#'   * <https://developers.google.com/chart/interactive/docs/spreadsheets#queryurlformat>
+#'
+#' Examples of datasource URLs
+#' ```
+#' # here is the base URL:
+#' https://docs.google.com/spreadsheets/d/SPREADSHEET_ID/gviz/tq
+#'
+#' # examples of the query part
+#' ?tqx=out:csv&sheet={sheet_name}
+#' ?range=A1:C4
+#' ?headers=N
+#' ?gid=N
+#' ````
+#'
+#' A readr issue that demonstrates sending an Authorization header
+#'   * <https://github.com/tidyverse/readr/issues/935>
+#'
+#' @inheritParams sheets_cells
+#' @param .url Whether to use the `"export"` (default) or `"datasource"` URL.
+#' @param .method Whether to process the CSV data with `vroom::vroom()`
+#'   (default) or `readr::read_csv()`.
+#' @param ... Passed along to `vroom::vroom()` or `readr::read_csv()`.
+#'
+#' @return A [tibble][tibble::tibble-package]
+#' @export
+#'
+#' @examples
+#' # Using the large NBA sheet from
+#' # https://github.com/tidyverse/googlesheets4/issues/122
+#' spreadsheet_id <- "1mnWcn7bd7obaXd05rnXrEtgzMBLdy7ctsYvlQM52W00"
+#' (ss <- as_sheets_id(spreadsheet_id))
+#'
+#' sheets_speedread(ss, .url = "export", .method = "vroom")
+#' sheets_speedread(ss, .url = "export", .method = "readr")
+#' sheets_speedread(ss, .url = "datasource", .method = "vroom")
+#' sheets_speedread(ss, .url = "datasource", .method = "readr")
+#'
+#' if (FALSE) {
+#' library(bench)
+#' bnch <- bench::mark(
+#'   sheets_speedread(ss, .url = "export", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "export", .method = "readr"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "vroom"),
+#'   sheets_speedread(ss, .url = "datasource", .method = "readr"),
+#'   iterations = 5,
+#'   check = FALSE
+#' )
+#' bnch
+#' }
+#'
+#' # prove that we can send a cell range, sheet, and col spec through ...
+#' (df <- sheets_speedread(
+#'   sheets_example("deaths"),
+#'   sheet = "other",
+#'   range = "A5:F15",
+#'   col_types = vroom::cols(
+#'     Age = vroom::col_integer(),
+#'     `Date of birth` = vroom::col_date("%m/%d/%Y"),
+#'     `Date of death` = vroom::col_date("%m/%d/%Y")
+#'   ),
+#'   .method = "vroom"
+#' ))
+#'
```

Comment:
@jimhester You should see a couple of things here.

First, notice how the `0s` leftover from vroom's progress reporter causes the variable type info to be misaligned.

```
> (df <- sheets_speedread(
+   sheets_example("deaths"),
+   sheet = "other",
+   range = "A5:F15",
+   col_types = vroom::cols(
+     Age = vroom::col_integer(),
+     `Date of birth` = vroom::col_date("%m/%d/%Y"),
+     `Date of death` = vroom::col_date("%m/%d/%Y")
+   ),
+   .method = "vroom"
+ ))
Reading from "deaths", "other" sheet, range "A5:F15"
# A tibble: 10 x 6                                                                
   Name           Profession   Age `Has kids` `Date of birth` `Date of death`
    0s<chr>          <chr>      <int> <lgl>      <date>          <date>         
 1 Vera Rubin     scientist     88 TRUE       1928-07-23      2016-12-25     
 2 Mohamed Ali    athlete       74 TRUE       1942-01-17      2016-06-03     
 3 Morley Safer   journalist    84 TRUE       1931-11-08      2016-05-19     
 4 Fidel Castro   politician    90 TRUE       1926-08-13      2016-11-25     
 5 Antonin Scalia lawyer        79 TRUE       1936-03-11      2016-02-13     
 6 Jo Cox         politician    41 TRUE       1974-06-22      2016-06-16     
 7 Janet Reno     lawyer        78 FALSE      1938-07-21      2016-11-07     
 8 Gwen Ifill     journalist    61 FALSE      1955-09-29      2016-11-14     
 9 John Glenn     astronaut     95 TRUE       1921-07-28      2016-12-08     
10 Pat Summit     coach         64 TRUE       1952-06-14      NA        
```

Second, why is that last `NA` there for Pat Summit's date of death? I don't see it with the readr method.

```
> (df <- sheets_speedread(
+   sheets_example("deaths"),
+   sheet = "other",
+   range = "A5:F15",
+   col_types = readr::cols(
+     Age = readr::col_integer(),
+     `Date of birth` = readr::col_date("%m/%d/%Y"),
+     `Date of death` = readr::col_date("%m/%d/%Y")
+   ),
+   .method = "readr"
+ ))
Reading from "deaths", "other" sheet, range "A5:F15"
# A tibble: 10 x 6
   Name           Profession   Age `Has kids` `Date of birth` `Date of death`
   <chr>          <chr>      <int> <lgl>      <date>          <date>         
 1 Vera Rubin     scientist     88 TRUE       1928-07-23      2016-12-25     
 2 Mohamed Ali    athlete       74 TRUE       1942-01-17      2016-06-03     
 3 Morley Safer   journalist    84 TRUE       1931-11-08      2016-05-19     
 4 Fidel Castro   politician    90 TRUE       1926-08-13      2016-11-25     
 5 Antonin Scalia lawyer        79 TRUE       1936-03-11      2016-02-13     
 6 Jo Cox         politician    41 TRUE       1974-06-22      2016-06-16     
 7 Janet Reno     lawyer        78 FALSE      1938-07-21      2016-11-07     
 8 Gwen Ifill     journalist    61 FALSE      1955-09-29      2016-11-14     
 9 John Glenn     astronaut     95 TRUE       1921-07-28      2016-12-08     
10 Pat Summit     coach         64 TRUE       1952-06-14      2016-06-28    
```

Reply:
If there is a `0s` "left over" it is likely a RStudio bug, did you try running this outside of RStudio?

Reply:
> did you try running this outside of RStudio?

I have now and your are correct. No leftover progress. This also only affects a very narrow situation so I shall not worry about it.